stoppable_workers_test.go

```diff
@@ -0,0 +1,75 @@
+package utils_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"go.viam.com/test"
+
+	"go.viam.com/utils"
+)
+
+func TestStoppableWorkers(t *testing.T) {
+	// Goleak checks from `VerifyTestMain` for `utils_test` should cause the
+	// following tests to fail if `StoppableWorkers` leaks any goroutines.
+	ctx := context.Background()
+
+	t.Run("one worker", func(t *testing.T) {
+		sw := utils.NewStoppableWorkers(ctx)
+		test.That(t, sw.Add(normalWorker), test.ShouldBeNil)
+		sw.Stop()
+	})
+
+	t.Run("concurrent workers", func(t *testing.T) {
+		sw := utils.NewStoppableWorkers(ctx)
+		concurrentWorker := func(ctx context.Context) {
+			go normalWorker(ctx)
+		}
+		test.That(t, sw.Add(concurrentWorker), test.ShouldBeNil)
+		test.That(t, sw.Add(concurrentWorker), test.ShouldBeNil)
+		sw.Stop()
+	})
+
+	t.Run("nested workers", func(t *testing.T) {
+		sw := utils.NewStoppableWorkers(ctx)
+		test.That(t, sw.Add(nestedWorkersWorker), test.ShouldBeNil)
+		sw.Stop()
+	})
+
+	t.Run("panicking worker", func(t *testing.T) {
+		sw := utils.NewStoppableWorkers(ctx)
+		// Both adding and stopping a panicking worker should cause no `panic`s.
+		test.That(t, sw.Add(panickingWorker), test.ShouldBeNil)
+		sw.Stop()
+	})
+
+	t.Run("already stopped", func(t *testing.T) {
+		sw := utils.NewStoppableWorkers(ctx)
+		sw.Stop()
+		test.That(t, sw.Add(normalWorker), test.ShouldBeError,
+			utils.ErrStoppableWorkersAlreadyStopped)
+		sw.Stop() // stopping twice should cause no `panic`
+	})
+}
+
+func normalWorker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(10 * time.Millisecond):
+		}
+	}
+}
```

Comment:
Could we add tests where the worker is doing something a bit more active too? We've got workers reading and writing from serial/i2c/spi protocols, workers returning immediately after math.

I'm also worried about the done context argument being passed it, it was an excellent safety rail for resource construction and not passing in a context with a cancel or timeout you don't know about.

Reply:
> Could we add tests where the worker is doing something a bit more active too?

Yep!

> I'm also worried about the done context argument being passed it

Slightly confused what you're referencing here. That workers get passed the `StoppableWorkers` context?

Reply:
hmm, maybe also a test that adds a few workers

Reply:
> Slightly confused what you're referencing here. That workers get passed the StoppableWorkers context?

No, that's fine and expected; I was also worried about the context being passed into `NewStoppableWorkers`, but you're making another creator function so I won't be anymore.